### PR TITLE
[WIP] Adding shared HPO logic

### PR DESCRIPTION
**Pull request by shagunsodhani**
## Motivation

The different HPO plugins (Ax, Nevergrad) can share some parsing logic

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

WIP

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

https://github.com/facebookresearch/hydra/issues/437

### CODE REVIEW

**Comment by shagunsodhani** [comment]
Sharing for some early feedback

File: hydra/core/override_parser/hpo.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ValueType = Union[str, int, float]
+
+
+@dataclass
+class HPOParameter:
+    name: str
+    type: str
+    log: bool
+    values: List[ValueType]
+    tags: List[str]
```
Comment (shagunsodhani) [inline hydra/core/override_parser/hpo.py:16]: Not all tag are common across the plugins. For example, NG uses `ordered` tag. We can either make that an attribute of the class or keep that in `tags`

File: hydra/core/override_parser/hpo.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        elif override.is_range_sweep():
+            hpo_type = "range"
+            values = [val for val in override.sweep_iterator()]
+        elif override.is_interval_sweep():
+            hpo_type = "interval"
+            value = override.value()
+            assert isinstance(value, IntervalSweep)
+            values = [value.start, value.end]
+    elif not override.is_hydra_override():
+        hpo_type = "fixed"
```
Comment (shagunsodhani) [inline hydra/core/override_parser/hpo.py:38]: This is the only new type I am introducing. Other types (choice, range..) are supported by Hydra override parser.

File: plugins/hydra_ax_sweeper/hydra_plugins/hydra_ax_sweeper/_core.py [hunk 3]
```diff
@@ -265,45 +271,30 @@ def chunks(batch: List[Any], n: Optional[int]) -> Iterable[List[Any]]:
...
 
-def create_fixed_param_from_element_override(override: Override) -> Dict[str, Any]:
-    key = override.get_key_element()
+def create_fixed_param(hpo_parameter: HPOParameter) -> Dict[str, Any]:
     param = {
-        "name": key,
+        "name": hpo_parameter.name,
         "type": "fixed",
-        "value": override.value(),
+        "value": hpo_parameter.values[0],
```
Comment (shagunsodhani) [inline plugins/hydra_ax_sweeper/hydra_plugins/hydra_ax_sweeper/_core.py:297]: In this case, values is a list of size 1

File: plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/_impl.py [hunk 4]
```diff
@@ -56,31 +55,33 @@ def create_nevergrad_param_from_config(
...
-        return ng.p.Choice(vals)
-    elif override.is_interval_sweep():
-        assert isinstance(val, IntervalSweep)
-        if "log" in val.tags:
-            scalar = ng.p.Log(lower=val.start, upper=val.end)
+            return ng.p.Choice(hpo_parameter.values)
+    elif hpo_parameter.type == "range":
+        return ng.p.Choice(hpo_parameter.values)
+    elif hpo_parameter.type == "interval":
+        lower = hpo_parameter.values[0]
```
Comment (shagunsodhani) [inline plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/_impl.py:69]: `lower` and `upper` can be made attributes of `HPOParameter` class

**Comment by shagunsodhani** [comment]
Bunch of lint errors - will fix them in the future iteration :)

**Comment by omry** [comment]
This is not what I had in mind, let's talk.

**Comment by omry** [comment]
Closing as this is not what I had in mind and we will eventually have proper implementation for this.

**Pull request closed without merging**